File: app/src/main/java/com/gproust/sprout/ui/sleep/SleepScreen.kt
package com.gproust.sprout.ui.sleep

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bedtime
import androidx.compose.material3.Button
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.gproust.sprout.R
import com.gproust.sprout.data.SproutRepository
import com.gproust.sprout.data.local.SleepEntity
import com.gproust.sprout.ui.common.AddEntryFab
import com.gproust.sprout.ui.common.AddEntrySheet
import com.gproust.sprout.ui.common.ConfirmDeleteDialog
import com.gproust.sprout.ui.common.DatePickerField
import com.gproust.sprout.ui.common.DayHeader
import com.gproust.sprout.ui.common.EmptyHint
import com.gproust.sprout.ui.common.EntryCard
import com.gproust.sprout.ui.common.FieldLabel
import com.gproust.sprout.ui.common.NotesField
import com.gproust.sprout.ui.common.SproutTopBar
import com.gproust.sprout.ui.common.TimePickerField
import com.gproust.sprout.ui.common.formatDuration
import com.gproust.sprout.ui.common.formatTime
import com.gproust.sprout.ui.common.nextDay
import com.gproust.sprout.ui.common.startOfDay
import com.gproust.sprout.ui.rememberSproutViewModelFactory
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class SleepViewModel(private val repository: SproutRepository) : ViewModel() {
    val sleeps = repository.sleeps
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun add(entity: SleepEntity) = viewModelScope.launch { repository.addSleep(entity) }
    fun delete(entity: SleepEntity) = viewModelScope.launch { repository.deleteSleep(entity) }
}

@Composable
fun SleepScreen() {
    val vm: SleepViewModel = viewModel(factory = rememberSproutViewModelFactory())
    val sleeps by vm.sleeps.collectAsState()
    val context = LocalContext.current

    var adding by remember { mutableStateOf(false) }
    var deleting by remember { mutableStateOf<SleepEntity?>(null) }

    if (adding) {
        AddEntrySheet(
            title = stringResource(R.string.sleep_log_title),
            onDismiss = { adding = false },
        ) {
            SleepForm(onAdd = { vm.add(it); adding = false })
        }
    }

    deleting?.let { entry ->
        ConfirmDeleteDialog(
            onConfirm = { vm.delete(entry); deleting = null },
            onDismiss = { deleting = null },
        )
    }

    val byDay = remember(sleeps) { sleeps.groupBy { startOfDay(it.startTime) } }

    Scaffold(
        topBar = { SproutTopBar(stringResource(R.string.screen_sleep)) },
        floatingActionButton = {
            AddEntryFab(stringResource(R.string.sleep_log_title)) { adding = true }
        },
    ) { padding ->
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(padding),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            if (sleeps.isEmpty()) {
                item { EmptyHint(stringResource(R.string.sleep_empty)) }
            }
            byDay.forEach { (day, entries) ->
                item(key = "day-$day") { DayHeader(day) }
                items(entries, key = { it.id }) { entry ->
                    val subtitle = entry.endTime?.let {
                        stringResource(
                            R.string.sleep_range,
                            formatTime(entry.startTime),
                            formatTime(it),
                            formatDuration(context, it - entry.startTime),
                        )
                    } ?: stringResource(R.string.sleep_ongoing, formatTime(entry.startTime))
                    EntryCard(
                        title = stringResource(R.string.sleep_entry_title),
                        subtitle = if (entry.notes.isNullOrBlank()) subtitle else "$subtitle\n${entry.notes}",
                        meta = formatTime(entry.startTime),
                        icon = Icons.Filled.Bedtime,
                        onDelete = { deleting = entry },
                    )
                }
            }
        }
    }
}

@Composable
private fun SleepForm(onAdd: (SleepEntity) -> Unit) {
    var start by remember { mutableLongStateOf(System.currentTimeMillis()) }
    var end by remember { mutableLongStateOf(System.currentTimeMillis()) }
    var hasEnded by remember { mutableStateOf(true) }
    var notes by remember { mutableStateOf("") }

    // The night a sleep started. Naps and nights alike are often logged later,
    // so the date is pickable, starting on today; a future one can't have
    // happened yet.
    FieldLabel(stringResource(R.string.field_date))
    DatePickerField(
        label = stringResource(R.string.picker_on),
        millis = start,
        allowFuture = false,
        onChange = { picked ->
            // Keep the wake time on the same day as the bedtime it was entered
            // against, so only the day moves.
            val shift = picked - start
            start = picked
            end += shift
        },
    )

    FieldLabel(stringResource(R.string.field_start))
    TimePickerField(label = stringResource(R.string.picker_from), millis = start, onChange = { start = it })

    Row(
        Modifier.fillMaxWidth().padding(top = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(stringResource(R.string.sleep_already_woke))
        Switch(checked = hasEnded, onCheckedChange = { hasEnded = it })
    }

    if (hasEnded) {
        FieldLabel(stringResource(R.string.field_end))
        TimePickerField(label = stringResource(R.string.picker_to), millis = end, onChange = { end = it })
    }

    Spacer(Modifier.height(8.dp))
    NotesField(value = notes, onChange = { notes = it })

    Spacer(Modifier.height(12.dp))
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
        Button(onClick = {
            onAdd(
                SleepEntity(
                    startTime = start,
                    // Both times are picked on the start's day, so a wake time
                    // before the bedtime means the sleep ran past midnight.
                    endTime = if (hasEnded) (if (end < start) nextDay(end) else end) else null,
                    notes = notes.ifBlank { null },
                ),
            )
        }) {
            Text(stringResource(R.string.sleep_add))
        }
    }
}

File: ios/Sprout/Sources/Screens/Sleep/SleepScreen.swift
import SproutData
import SwiftUI

/// Naps and nights, from `ui/sleep/SleepScreen.kt`.
@Observable
@MainActor
final class SleepViewModel {
    var sleeps: [Sleep] = []

    private let repository: SproutRepository

    init(repository: SproutRepository) {
        self.repository = repository
    }

    func observeSleeps() async {
        await observe(repository.sleeps) { [weak self] in self?.sleeps = $0 }
    }

    func add(_ sleep: Sleep) {
        try? repository.addSleep(sleep)
    }

    func delete(_ sleep: Sleep) {
        try? repository.deleteSleep(sleep)
    }

    /// Close a sleep that was logged as still running.
    ///
    /// Without this the only way out of an ongoing entry is to delete it and
    /// type it again — and while it stays open it goes on counting, because the
    /// dashboard and the statistics both read a missing end as "still asleep".
    func wake(_ sleep: Sleep) {
        var updated = sleep
        updated.endTime = Clock.millis
        try? repository.updateSleep(updated)
    }

    /// Entries grouped by the day they started, newest day first.
    var byDay: [(day: Int64, entries: [Sleep])] {
        Dictionary(grouping: sleeps) { SproutFormat.startOfDay($0.startTime) }
            .map { (day: $0.key, entries: $0.value) }
            .sorted { $0.day > $1.day }
    }
}

struct SleepScreen: View {
    @Environment(\.sprout) private var sprout
    @State private var model: SleepViewModel?
    @State private var adding = false
    @State private var pendingDelete: Sleep?

    var body: some View {
        Group {
            if let model {
                content(model)
            } else {
                Color.clear
            }
        }
        .navigationTitle(Str.t("screen_sleep"))
        .navigationBarTitleDisplayMode(.inline)
        .task {
            let model = model ?? SleepViewModel(repository: sprout.repository)
            self.model = model
            await model.observeSleeps()
        }
    }

    @ViewBuilder
    private func content(_ model: SleepViewModel) -> some View {
        ZStack(alignment: .bottomTrailing) {
            ScrollView {
                LazyVStack(alignment: .leading, spacing: Spacing.snug) {
                    if model.sleeps.isEmpty {
                        EmptyHint(Str.t("sleep_empty"))
                    }
                    ForEach(model.byDay, id: \.day) { group in
                        DayHeader(dayStartMillis: group.day, now: Clock.millis)
                        ForEach(group.entries) { entry in
                            card(entry, model: model)
                        }
                    }
                }
                .padding(Spacing.regular)
                // Clear of the add button, which floats over the list.
                .padding(.bottom, 80)
            }

            AddEntryButton(accessibilityLabel: Str.t("sleep_log_title")) { adding = true }
                .padding(Spacing.regular)
        }
        .sproutStyle()
        .sheet(isPresented: $adding) {
            SleepForm { model.add($0); adding = false }
        }
        .confirmDelete(
            isPresented: Binding(get: { pendingDelete != nil }, set: { if !$0 { pendingDelete = nil } })
        ) {
            if let entry = pendingDelete { model.delete(entry) }
            pendingDelete = nil
        }
    }

    private func card(_ entry: Sleep, model: SleepViewModel) -> some View {
        EntryCard(
            title: Str.t("sleep_entry_title"),
            subtitle: subtitle(entry),
            meta: SproutDateStyle.time(entry.startTime),
            systemImage: "moon.zzz.fill",
            onDelete: { pendingDelete = entry },
            details: { EmptyView() },
            action: {
                // A sleep still running is the one entry with something left to
                // say; ending it is one tap, from here.
                if entry.isOngoing {
                    Button(Str.t("sleep_woke_up")) { model.wake(entry) }
                        .font(.footnote)
                        .foregroundStyle(SproutColor.primary)
                }
            }
        )
    }

    /// When and how long, then where and how they were lying if that was noted,
    /// then the note itself — one line each.
    private func subtitle(_ entry: Sleep) -> String {
        var lines: [String] = []

        if let end = entry.endTime {
            lines.append(
                Str.t(
                    "sleep_range",
                    SproutDateStyle.time(entry.startTime),
                    SproutDateStyle.time(end),
                    SproutFormat.duration(millis: end - entry.startTime).text
                )
            )
        } else {
            lines.append(Str.t("sleep_ongoing", SproutDateStyle.time(entry.startTime)))
        }

        let details = entry.details
        if !details.isEmpty { lines.append(details) }
        if let notes = entry.notes?.trimmingCharacters(in: .whitespacesAndNewlines), !notes.isEmpty {
            lines.append(notes)
        }
        return lines.joined(separator: "\n")
    }
}

extension Sleep {
    /// Where this sleep happened — the parent's own word for it when they gave
    /// one, since a list of six cannot cover where a baby actually falls asleep.
    var placeLabel: String? {
        guard let place else { return nil }
        let named = placeNote?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return place == .OTHER && !named.isEmpty ? named : place.label
    }

    /// "Their own bed · On their back" — whichever of the two was recorded.
    var details: String {
        [placeLabel, position?.label]
            .compactMap { $0 }
            .joined(separator: Str.t("sleep_details_separator"))
    }
}

// MARK: - The form

private struct SleepForm: View {
    let onAdd: (Sleep) -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var start = Clock.millis
    @State private var end = Clock.millis
    @State private var hasEnded = true
    @State private var position: SleepPosition?
    @State private var place: SleepPlace?
    @State private var placeNote = ""
    @State private var notes = ""

    var body: some View {
        NavigationStack {
            Form {
                Section {
                    // Naps and nights alike are often logged later, so both
                    // times are pickable. The day moves them together.
                    DateTimeField(label: Str.t("field_start"), millis: $start)
                    Toggle(Str.t("sleep_already_woke"), isOn: $hasEnded)
                    if hasEnded {
                        DateTimeField(label: Str.t("field_end"), millis: $end)
                    }
                }

                Section {
                    FieldLabel(Str.t("sleep_place_label"))
                    // Optional, and not asked twice: tapping the chosen chip
                    // again clears it, so a sleep is still two taps (BDR-0014).
                    ChoiceChips(
                        options: SleepPlace.allCases,
                        selection: $place,
                        label: \.label,
                        allowsDeselection: true
                    )
                    .onChange(of: place) { _, new in
                        if new != .OTHER { placeNote = "" }
                    }
                    if place == .OTHER {
                        TextField(Str.t("sleep_place_other_hint"), text: $placeNote)
                    }

                    FieldLabel(Str.t("sleep_position_label"))
                    ChoiceChips(
                        options: SleepPosition.allCases,
                        selection: $position,
                        label: \.label,
                        allowsDeselection: true
                    )
                }

                Section {
                    NotesField(text: $notes)
                }
            }
            .navigationTitle(Str.t("sleep_log_title"))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(Str.t("action_cancel")) { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button(Str.t("sleep_add")) { onAdd(built) }
                }
            }
        }
    }

    private var built: Sleep {
        Sleep(
            startTime: start,
            // Both times are picked on the start's day, so a wake time before
            // the bedtime means the sleep ran past midnight.
            endTime: hasEnded ? (end < start ? SproutFormat.nextDay(end) : end) : nil,
            position: position,
            place: place,
            // Only "somewhere else" carries a name; the other five are named by
            // the app, in whatever language it is showing.
            placeNote: place == .OTHER
                ? placeNote.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
                : nil,
            notes: notes.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        )
    }
}

extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}
